src/Response.php

```diff
@@ -77,34 +77,48 @@ public function getHeaders()
 
     public function handleData($data)
     {
-        $this->emit('data', array($data, $this));
+        if ($this->readable) {
+            $this->emit('data', array($data));
+        }
     }
 
     public function handleEnd()
     {
+        if (!$this->readable) {
+            return;
+        }
+        $this->emit('end', array());
         $this->close();
     }
 
     public function handleError(\Exception $error)
     {
+        if (!$this->readable) {
+            return;
+        }
         $this->emit('error', array(new \RuntimeException(
             "An error occurred in the underlying stream",
             0,
             $error
-        ), $this));
+        )));
+
+        $this->close();
+    }
 
-        $this->close($error);
+    public function handleClose()
```

Comment:
Should probably be marked `@internal` (see also others?) or we may even use `close()` as the event handler here.